File: src/0103.binary-tree-zigzag-level-order-traversal.103/solution.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
// 层序遍历
var zigzagLevelOrder = function (root) {
    const result = [];
    if (root === null) {
        return result;
    }

    let curLevel = [
        root, ];
    while (true) {
        const nextLevel = [];
        const values = [];
        for (let i = 0; i < curLevel.length; i++) {
            values.push(curLevel[i].val);
            curLevel[i].left && nextLevel.push(curLevel[i].left);
            curLevel[i].right && nextLevel.push(curLevel[i].right);
        }
        result.length % 2 === 1 && values.reverse();
        result.push(values);
        if (nextLevel.length === 0) {
            return result;
        }
        curLevel = nextLevel;
    }
};

File: src/0103.binary-tree-zigzag-level-order-traversal.103/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class Solution {
    public List<List<Integer>> zigzagLevelOrder(TreeNode root) {
        if (root == null) {
            return List.of();
        }
        List<List<Integer>> result = new ArrayList<>();

        travel(List.of(root), result, false);

        return result;
    }

    private void travel(List<TreeNode> level, List<List<Integer>> result, boolean reverse) {
        List<Integer> row = new ArrayList<>();
        List<TreeNode> nextLevel = new ArrayList<>();
        for (TreeNode node : level) {
            row.add(node.val);
            if (node.left != null) {
                nextLevel.add(node.left);
            }
            if (node.right != null) {
                nextLevel.add(node.right);
            }

        }
        if (reverse) {
            row = row.reversed();
        }
        result.add(row);

        if (!nextLevel.isEmpty()) {
            travel(nextLevel, result, !reverse);
        }

    }
}